### PR TITLE
[SPARK-16318][SQL] Implement all remaining xpath functions

### PR DESCRIPTION
**Pull request by petermaxlee**
## What changes were proposed in this pull request?

This patch implements all remaining xpath functions that Hive supports and not natively supported in Spark: xpath_int, xpath_short, xpath_long, xpath_float, xpath_double, xpath_string, and xpath.
## How was this patch tested?

Added unit tests and end-to-end tests.

### CODE REVIEW

**Comment by petermaxlee** [comment]
Can you guys take a look at this? @cloud-fan @dongjoon-hyun @gatorsmile

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/xml/XPathExpressionSuite.scala [hunk 1]
```diff
@@ -27,33 +27,127 @@ import org.apache.spark.sql.types.StringType
...
+    testExpr("<a><b>bb</b><c>cc</c></a>", "a/b", "bb")
+    testExpr("<a><b>bb</b><c>cc</c></a>", "a/c", "cc")
+    testExpr("<a><b>bb</b><c>cc</c></a>", "a/d", "")
+    testExpr("<a><b>b1</b><b>b2</b></a>", "//b", "b1")
+    testExpr("<a><b>b1</b><b>b2</b></a>", "a/b[1]", "b1")
+    testExpr("<a><b>b1</b><b id='b_2'>b2</b></a>", "a/b[@id='b_2']", "b2")
+  }
+
+  test("null handling") {
+    // We only do this for one expression since they all share the same common base implementation
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/xml/XPathExpressionSuite.scala]: seems the boolean and string one doesn't share same implementation?(`xpathUtil.evalBoolean` and `xpathUtil.evalString`)

**Comment by gatorsmile** [comment]
Could you add some test cases for the boundary values? That means, trying the maximum and minimum values for each different data type. Also, the values that are out of the boundary.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/xml/XPathExpressionSuite.scala [hunk 2]
```diff
@@ -27,33 +27,127 @@ import org.apache.spark.sql.types.StringType
...
-    testBoolean("<a><b>10</b></a>", "a/b = 10", true)
-
-    // null input
-    testBoolean(null, null, null)
-    testBoolean(null, "a", null)
-    testBoolean("<a><b>10</b></a>", null, null)
-
-    // exception handling for invalid input
+  test("invalid xml handling") {
     intercept[Exception] {
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/xml/XPathExpressionSuite.scala]: Which Exception will we throw? What is the message we issue? When users see the message, can they understand the cause?

**Comment by gatorsmile** [comment]
Could you add more test cases for corrupted xml and illegal inputs? So far, the test suite `XPathExpressionSuite` only covers the basic cases.

**Comment by gatorsmile** [comment]
Normally, for each function or data type we implement in an enterprise product, we need to document the limit/restriction. That is why I am asking for the boundary values, negative cases and error handling.

**Comment by cloud-fan** [comment]
+1 for @gatorsmile 's advices

**Comment by petermaxlee** [comment]
Thanks for the feedback. I will update this later.

**Comment by rxin** [comment]
For this one I think we should consider supporting only foldable literals for the path component.

**Comment by rxin** [comment]
Also - rather than having concrete implementations for all of these, why don't we use RuntimeReplaceable?

**Comment by petermaxlee** [comment]
I pushed a new change to this. We now have better error messages and test coverage for those. These expressions also now require foldable paths.

I also changed the test values to make sure we are not restricting the range of values an expression can return (e.g. for XPathInt the test case makes sure it can return a value larger than short). However, I don't think it'd make sense to test overflow behavior because those are mostly undefined throughout Spark SQL.

Re: @rxin's suggestion on using RuntimeReplaceable -- I did try that, but I'm afraid the current implementation of RuntimeReplaceable isn't meant for stacking multiple expressions together, and time checking actually breaks if XPathInt replaces itself with Cast(XPathDouble). I looked into fixing that, but it seemed more complicated and it doesn't actually save much space (it actually takes more code to use RuntimeReplaceable).

**Comment by petermaxlee** [comment]
I just added the general xpath function that returns an array of string too.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by petermaxlee** [comment]
@cloud-fan Jenkins already ran twice successfully before.

**Comment by cloud-fan** [comment]
The latest result is `This patch does not merge cleanly.`, I just wanna double check it.

**Comment by petermaxlee** [comment]
I guess whatever generates that message is buggy?

**Comment by yhuai** [comment]
As a follow-up task. Can you take a look at the following query files and add useful tests in your test? Thanks.
```
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/describe_xpath.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/input_testxpath.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/input_testxpath2.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/input_testxpath3.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/input_testxpath4.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/udf_xpath.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/udf_xpath_boolean.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/udf_xpath_double.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/udf_xpath_float.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/udf_xpath_int.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/udf_xpath_long.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/udf_xpath_short.q
.//sql/hive/src/test/resources/ql/src/test/queries/clientpositive/udf_xpath_string.q
```

**Comment by petermaxlee** [comment]
Actually I created the unit tests based on those.

**Comment by petermaxlee** [comment]
BTW I have to say Hive's test coverage in this area is very spotty, so I don't actually think it's great to follow, but I used those.

**Comment by yhuai** [comment]
OK. Thanks. Then, it will be good to add more tests for cases that are not covered by those hive tests.

**Comment by cloud-fan** [comment]
Thanks, merging to master! This doesn't merge clearly to 2.0, @petermaxlee can you submit a new PR against 2.0? thanks!

**Comment by petermaxlee** [comment]
@cloud-fan thanks for merging!

@yhuai I think the degree to which we want to add more tests also depend on how much we trust the library we are using. XPath (with Query) is almost as complicated as SQL itself.

**Comment by petermaxlee** [comment]
Here it is for branch-2.0.
https://github.com/apache/spark/pull/14131

**Pull request closed without merging**